tiatoolbox/models/engine/io_config.py

```diff
@@ -0,0 +1,421 @@
+        highest_input_resolution (dict):
+            Highest resolution to process the image based on input and
+            output resolutions. This helps to read the image at the optimal
+            resolution and improves performance.
+
+    Examples:
+        >>> # Defining io for a base network and converting to baseline.
+        >>> ioconfig = ModelIOConfigABC(
+        ...     input_resolutions=[{"units": "mpp", "resolution": 0.5}],
+        ...     output_resolutions=[{"units": "mpp", "resolution": 1.0}],
+        ...     patch_input_shape=(224, 224),
+        ...     stride_shape=(224, 224),
+        ... )
+        >>> ioconfig = ioconfig.to_baseline()
+
+    """
+
+    input_resolutions: list[dict]
+    patch_input_shape: list[int] | np.ndarray | tuple[int, int]
+    stride_shape: list[int] | np.ndarray | tuple[int, int] = None
+    output_resolutions: list[dict] = field(default_factory=list)
+
+    def __post_init__(self):
+        """Perform post initialization tasks."""
+        if self.stride_shape is None:
+            self.stride_shape = self.patch_input_shape
+
+        self.resolution_unit = self.input_resolutions[0]["units"]
+
+        if self.resolution_unit == "mpp":
+            self.highest_input_resolution = min(
+                self.input_resolutions,
+                key=lambda x: x["resolution"],
+            )
+        else:
+            self.highest_input_resolution = max(
+                self.input_resolutions,
+                key=lambda x: x["resolution"],
+            )
+
+        self._validate()
+
+    def _validate(self):
+        """Validate the data format."""
+        resolutions = self.input_resolutions + self.output_resolutions
+        units = [v["units"] for v in resolutions]
+        units = np.unique(units)
+
+        if len(units) != 1:
+            msg = (
+                f"Multiple resolution units found: `{units}`. "
+                f"Mixing resolution units is not allowed."
+            )
+            raise ValueError(
+                msg,
+            )
+
+        if units[0] not in [
+            "power",
+            "baseline",
+            "mpp",
+        ]:
+            msg = f"Invalid resolution units `{units[0]}`."
+            raise ValueError(msg)
+
+    @staticmethod
+    def scale_to_highest(resolutions: list[dict], units: Units):
+        """Get the scaling factor from input resolutions.
+
+        This will convert resolutions to a scaling factor with respect to
+        the highest resolution found in the input resolutions list.
+
+        Args:
+            resolutions (list):
+                A list of resolutions where one is defined as
+                `{'resolution': value, 'unit': value}`
+            units (Units):
+                Resolution units.
+
+        Returns:
+            :class:`numpy.ndarray`:
+                A 1D array of scaling factors having the same length as
+                `resolutions`
+
```

Comment:
Could we have an example here? I am slightly confused by the description. Also no return type hint.

Reply:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/618/commits/ab762f9a6b7eb247ebacb8a0c91a222e757967ab

Reply:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/618/commits/15887342c589da1eb9285de9feb4f0a24bab5251